ir/src/lib.rs

```diff
@@ -0,0 +1,203 @@
+                    air_name = air_def;
+                }
+                ast::SourceSection::TraceCols(columns) => {
+                    for (idx, Identifier(name)) in columns.main_cols.iter().enumerate() {
+                        trace_columns.insert(name, idx)?;
+                    }
+                }
+                _ => {}
+            }
+        }
+
+        // then process the constraints.
+        let mut main_boundary_constraints = BoundaryConstraints::default();
+        let mut main_transition_constraints = TransitionConstraints::default();
+        for section in source {
+            match section {
+                ast::SourceSection::BoundaryConstraints(constraints) => {
+                    for constraint in constraints.boundary_constraints.iter() {
+                        main_boundary_constraints.insert(constraint, &trace_columns)?;
+                    }
+                }
+                ast::SourceSection::TransitionConstraints(constraints) => {
+                    for constraint in constraints.transition_constraints.iter() {
+                        main_transition_constraints.insert(constraint, &trace_columns)?;
+                    }
+                }
+                _ => {}
+            }
+        }
+
+        Ok(Self {
+            air_name: air_name.to_string(),
+            main_boundary_constraints,
+            main_transition_constraints,
+        })
+    }
+
+    // --- PUBLIC ACCESSORS -----------------------------------------------------------------------
+    pub fn get_air_name(&self) -> &str {
+        &self.air_name
+    }
+
+    pub fn num_main_assertions(&self) -> usize {
+        self.main_boundary_constraints.len()
+    }
+
+    pub fn main_first_boundary_constraints(&self) -> Vec<&Expr> {
+        self.main_boundary_constraints.first()
+    }
+
+    pub fn main_last_boundary_constraints(&self) -> Vec<&Expr> {
+        self.main_boundary_constraints.last()
+    }
+
+    pub fn main_degrees(&self) -> Vec<u8> {
+        self.main_transition_constraints.degrees()
+    }
+
+    pub fn main_transition_constraints(&self) -> &[NodeIndex] {
+        self.main_transition_constraints.constraints()
+    }
+
+    pub fn main_transition_graph(&self) -> &AlgebraicGraph {
+        self.main_transition_constraints.graph()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use parser::{
+        grammar::SourceParser,
+        lexer::{Lexer, Token},
+    };
+
+    fn get_parsed(source: &str) -> ast::Source {
+        let lex = Lexer::new(source).spanned().map(Token::to_spanned);
+        SourceParser::new().parse(lex).unwrap()
+    }
+
+    #[test]
+    fn boundary_constraints() {
+        let source = "
+        trace_columns:
+            main: [clk]
+        boundary_constraints:
+            enf clk.first = 0
+            enf clk.last = 1";
+
+        let parsed = get_parsed(source);
+
+        let result = AirIR::from_source(&parsed);
+        assert!(result.is_ok());
```

Comment:
Were you planning to add checking for actual AirIR built later? Sorry if you already mentioned this somewhere and I missed it, just checking.

Reply:
Yes, we should definitely add more rigorous tests. I've added a TODO and opened #7